### PR TITLE
TASK: Use content repository clock for date time default value

### PR DESCRIPTION
**Pull request by mhsdesign**
We now use the content repository clock for date time default value.

That way we can deterministically test these as `$this->clock()->now()->modify("now")` is just the specified "now" in the tests:

```
And the current date and time is "2024-09-22T12:00:00+00:00"
```

Also all date time default values now respect the clock's timezone.

We still only accept a "string" or null for dates. Previously there were not so many constraint checks, but despite what `Symfony\Component\Serializer\Normalizer\DateTimeNormalizer` looks like it also only allowed "string"s.

Added tests:

- Test other possible php date time modifiers and also timezones
- Add test for `defaultValue: null` or `defaultValue: ~` works (relevant for inheritance)
- Add test for all illegal date time default values

**Upgrade instructions**

**Review instructions**

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the correct branch:
  - If it's a bugfix, use the [lowest maintained branch which has the bug](https://www.neos.io/features/release-roadmap.html)
  - If it's a non-breaking feature, use the branch of the next version (might be either minor or major)
  - If it's a breaking feature it should typically go into the next major version
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

### CODE REVIEW

**Comment by mhsdesign** [comment]
I targeted 9.2 because i intended to extract the `SerializedPropertyValues::defaultFromNodeType` some where else ... this is now a lot of ugly code onto a poor value object ... this should rather be a factory at this point with all the dependencies.

File: Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/02-NodeCreation/05-CreateNodeAggregateWithNode_ComplexDefaultAndInitialProperties.feature [hunk 1]
```diff
@@ -84,11 +106,30 @@ Feature: Create a node aggregate with complex default values
...
+      | parentNodeAggregateId | "lady-eleonode-rootford"              |
+    Then I expect a node identified by cs-identifier;nody-mc-nodeface;{} to exist in the content graph
+
+    And I expect this node to have the following properties:
+      | Key           | Value                                           |
+      | now           | Date:2024-09-22T12:00:00+02:00                  |
+      | tomorrow      | Date:2024-09-23T00:00:00+02:00                  |
+      | lastMonth     | Date:2024-08-22T12:00:00+02:00                  |
+      # Even though we specified the timestamp as ATOM with timezone in the defaults, PHPs date->modify() keeps the original time zone.
+      | date          | Date:2020-08-20T18:56:15+02:00                  |
```
Comment (mhsdesign) [inline Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/02-NodeCreation/05-CreateNodeAggregateWithNode_ComplexDefaultAndInitialProperties.feature]: Todo discuss if this is "oke". 

What happened here is that the default value

```
defaultValue: '2020-08-20T18:56:15+00:00'
```

is applied via `modify()` but unfortunately the timezone is lost. and now the one from the clock.

To fix this we would need to find out if the defaultValue is absolute and carries its own timezone, if so we would return it directly:

```php
$now = $clock->now();
$possiblyAbsoluteDateWithOwnTimeZone = new \DateTimeImmutable($defaultValue, $now->getTimezone());

if ($possiblyAbsoluteDateWithOwnTimeZone->getOffset() !== $now->getOffset()) {
    $deserializedDefaultValue = $possiblyAbsoluteDateWithOwnTimeZone;
} else {
    // relative date
    $deserializedDefaultValue = $now->modify($defaultValue);
}
```

Comment (mhsdesign) [reply Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/02-NodeCreation/05-CreateNodeAggregateWithNode_ComplexDefaultAndInitialProperties.feature]: I think introducing the extra complexity as outlined is not important now as i think no one will use an ATOM date default value WITH the timezone. If proven wrong we can change this.

But whats important with this PR that we use the actual clock at all which is a good and right step!

File: Neos.ContentRepository.Core/Classes/Feature/NodeModification/Dto/SerializedPropertyValues.php [hunk 2]
```diff
@@ -19,6 +19,7 @@
 use Neos\ContentRepository\Core\NodeType\NodeType;
 use Neos\ContentRepository\Core\SharedModel\Node\PropertyName;
 use Neos\ContentRepository\Core\SharedModel\Node\PropertyNames;
+use Psr\Clock\ClockInterface;
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/Feature/NodeModification/Dto/SerializedPropertyValues.php]: new PHP Namespace import nut not used?

File: Neos.ContentRepository.Core/Classes/Feature/NodeAggregateCommandHandler.php [hunk 3]
```diff
@@ -49,8 +49,10 @@
 use Neos\ContentRepository\Core\Feature\SubtreeTagging\Command\TagSubtree;
 use Neos\ContentRepository\Core\Feature\SubtreeTagging\Command\UntagSubtree;
 use Neos\ContentRepository\Core\Feature\SubtreeTagging\SubtreeTagging;
+use Neos\ContentRepository\Core\Infrastructure\Property\NodeTypeDefaultPropertySerializer;
 use Neos\ContentRepository\Core\Infrastructure\Property\PropertyConverter;
 use Neos\ContentRepository\Core\NodeType\NodeTypeManager;
+use Psr\Clock\ClockInterface;
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/Feature/NodeAggregateCommandHandler.php]: Again, I fail to understand why we need this namespace if it's not used in the new code!?

File: Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/02-NodeCreation/05-CreateNodeAggregateWithNode_ComplexDefaultAndInitialProperties.feature [hunk 4]
```diff
@@ -102,15 +143,108 @@ Feature: Create a node aggregate with complex default values
...
+            type: DateTimeImmutable
+            defaultValue: 'not a date'
+      """
     When the command CreateNodeAggregateWithNode is executed with payload and exceptions are caught:
       | Key                   | Value                                           |
       | nodeAggregateId       | "nody-mc-nodeface"                              |
       | nodeTypeName          | "Neos.ContentRepository.Testing:FaultyDateNode" |
       | parentNodeAggregateId | "lady-eleonode-rootford"                        |
-    And the last command should have thrown an exception of type "RuntimeException" with code 1708416598
+    And the last command should have thrown an exception of type "RuntimeException" with code 1783085627
```
Comment (bwaidelich) [inline Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/02-NodeCreation/05-CreateNodeAggregateWithNode_ComplexDefaultAndInitialProperties.feature:250]: Why did the exception code change?

File: Neos.ContentRepository.Core/Classes/Factory/ContentRepositoryFactory.php [hunk 5]
```diff
@@ -156,6 +157,10 @@ public function getOrBuild(): ContentRepository
                 $this->contentDimensionZookeeper,
                 $this->interDimensionalVariationGraph,
                 $this->propertyConverter,
+                new NodeTypeDefaultPropertySerializer(
+                    $this->propertyConverter,
+                    $this->clock,
+                )
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/Factory/ContentRepositoryFactory.php]: So now we need a separate serializer on top of the property converter? Can't we build the new functionality in to the former?

File: Neos.ContentRepository.Core/Classes/Feature/NodeAggregateCommandHandler.php [hunk 6]
```diff
@@ -137,11 +140,6 @@ protected function areAncestorNodeTypeConstraintChecksEnabled(): bool
         return $this->ancestorNodeTypeConstraintChecksEnabled;
     }
 
-    public function getPropertyConverter(): PropertyConverter
-    {
-        return $this->propertyConverter;
-    }
-
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/Feature/NodeAggregateCommandHandler.php:144]: Is it safe to delete this public method? and why is that related?

**Review by bwaidelich — COMMENTED** [review]
That's quite a lot of ceremony just to make relative dates in default values testable (if I got this right!?)

I struggle mostly with the fact that we introduce a new `NodeTypeDefaultPropertySerializer` that _also_ has a dependency on the `PropertyConverter`.
The cleaner cut would maybe a new class/interface that does both, converting property values and default values..
Or to resolve relative dates on a different layer maybe, it's not really a CR core feature IMO

File: Neos.ContentRepository.Core/Classes/Feature/NodeAggregateCommandHandler.php (hunk 3 above)
Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/Feature/NodeAggregateCommandHandler.php]: sorry a hickup because an intermediate solution looked like 

```php
$defaultValues = SerializedPropertyValues::defaultFromNodeType(
    $tetheredNodeType,
    $this->getPropertyConverter(),
    $this->clock
);
```

File: Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/02-NodeCreation/05-CreateNodeAggregateWithNode_ComplexDefaultAndInitialProperties.feature (hunk 4 above)
Comment (mhsdesign) [reply Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/02-NodeCreation/05-CreateNodeAggregateWithNode_ComplexDefaultAndInitialProperties.feature:250]: because we throw the exception now in the new code path in the `NodeTypeDefaultPropertySerializer` instead of from the PropertyConverter

File: Neos.ContentRepository.Core/Classes/Feature/NodeAggregateCommandHandler.php (hunk 6 above)
Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/Feature/NodeAggregateCommandHandler.php:144]: See https://github.com/neos/neos-development-collection/pull/5913/commits/c8bbe450f42603772afc440118104d643bdcb2c2 yes its not public as these are our internal command handlers where we do what we want. We have partly this pattern of abstract methods to share some dependencies for traits but this results in all dependencies everywhere without knowing why see commit and later commits which could remove this dependency in the structure adjustment

File: Neos.ContentRepository.Core/Classes/Factory/ContentRepositoryFactory.php (hunk 5 above)
Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/Factory/ContentRepositoryFactory.php]: i considered this as well but that would require the `Clock` then for the `PropertyConverter` which is also odd.

**Comment by mhsdesign** [comment]
> That's quite a lot of ceremony just to make relative dates in default values testable (if I got this right!?)

Jup thats why i targeted 9.2 because everything i tough kinda explodes.


> I struggle mostly with the fact that we introduce a new NodeTypeDefaultPropertySerializer that also has a dependency on the PropertyConverter.

I think anyway the decision is right to split it away from the `SerializedPropertyValues::defaultFromNodeType()` which is just overly bloated.

Also everything is internal and im not married with the code how it is now. I think the `PropertyConverter` is already wrongly cut as it seems to reside in the `Infrastructure` layer yet we invoke it from our Core CommandHandlers.

Im not sure where to stop engineering this pr. Maybe split this pr starting at https://github.com/neos/neos-development-collection/pull/5913/commits/e39bb07352e6c0c0b7ddba5644eb5f1a646c884e and just have the proper date handling already?

**Comment by mhsdesign** [comment]
Oke i decided that its enough change for the PR to keep it as it was before my new changes today introducing the `NodeTypeDefaultPropertySerializer` where i didnt hit your taste @bwaidelich immediately - and that is totally fine - as i agree there is still a code smell 

So lets keep it like this a little hacky but to be cleaned up later?

**Review by bwaidelich — APPROVED** [review]
I like this one much better, thanks for sticking through!

**Comment by mhsdesign** [comment]
Thanks a lot!

And here will the PR i might work on some day to turn this into an architectural masterpiece ... or not ... lets see:)

https://github.com/neos/neos-development-collection/pull/5935 

(its still on the state we discussed we yet dont like)
